[Bug] webpack@4.44.1 does not work with `node-modules` linker in a monorepo

**Describe the bug**

webpack@4.44.1 does not work with `node-modules` linker in a monorepo, the `webpack` is somehow NOT hoisted while its dependency `terser-webpack-plugin` with `webpack` as a `peerDependencies` is still hoisted.

**To Reproduce**

<details>
  <summary>Reproduction</summary>

  ```js repro
  require('fs').writeFileSync('.yarnrc.yml', 'nodeLinker: node-modules')

  await packageJson({ name: 'root', workspaces: ['packages/*'] })
  await packageJson(
    {
      name: 'package-a',
      devDependencies: {
        "webpack": "4.44.1",
        "webpack-cli": "3.3.12",
      },
    },
    { cwd: 'packages/package-a' },
  )
  await yarn('install', { maxBuffer: 64 * 1024 * 1024 })

  await expect(yarn(`node`, `packages/package-a/node_modules/.bin/webpack-cli`))
    .rejects.not.toThrow(`Cannot find module 'webpack/lib/RequestShortener'`)
  ```
</details>

**Screenshots**

```
Error: Cannot find module 'webpack/lib/RequestShortener'
Require stack:
- ~/node_modules/terser-webpack-plugin/dist/index.js
- ~/node_modules/terser-webpack-plugin/dist/cjs.js
- ~/packages/workspace-a/node_modules/webpack/lib/WebpackOptionsDefaulter.js
- ~/packages/workspace-a/node_modules/webpack/lib/webpack.js
- ~/packages/workspace-a/node_modules/webpack-cli/bin/utils/validate-options.js
- ~/packages/workspace-a/node_modules/webpack-cli/bin/utils/convert-argv.js
- ~/packages/workspace-a/node_modules/webpack-cli/bin/cli.js
```

**Environment**

 - Yarn version 2.1.1

**Additional context**

- `webpack@4.44.0` works fine.
- the issue is probably introduced by https://github.com/webpack/webpack/pull/11190, which adds peerDep `webpack-cli` to `webpack`
- there is a peerDep `webpack` in `webpack-cli`

--- a/.yarn/versions/c03daef8.yml
+++ b/.yarn/versions/c03daef8.yml
@@ -1,0 +1,25 @@
+releases:
+  "@yarnpkg/cli": patch
+  "@yarnpkg/plugin-node-modules": patch
+  "@yarnpkg/pnpify": patch
+
+declined:
+  - "@yarnpkg/plugin-compat"
+  - "@yarnpkg/plugin-constraints"
+  - "@yarnpkg/plugin-dlx"
+  - "@yarnpkg/plugin-essentials"
+  - "@yarnpkg/plugin-init"
+  - "@yarnpkg/plugin-interactive-tools"
+  - "@yarnpkg/plugin-npm-cli"
+  - "@yarnpkg/plugin-pack"
+  - "@yarnpkg/plugin-patch"
+  - "@yarnpkg/plugin-pnp"
+  - "@yarnpkg/plugin-stage"
+  - "@yarnpkg/plugin-typescript"
+  - "@yarnpkg/plugin-version"
+  - "@yarnpkg/plugin-workspace-tools"
+  - vscode-zipfs
+  - "@yarnpkg/builder"
+  - "@yarnpkg/core"
+  - "@yarnpkg/doctor"
+  - "@yarnpkg/pnp"

--- a/packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts
+++ b/packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts
@@ -1,4 +1,4 @@
-import {xfs, npath} from '@yarnpkg/fslib';
+import {xfs, npath, PortablePath} from '@yarnpkg/fslib';
 
 const {
   fs: {writeFile, writeJson},
@@ -349,6 +349,61 @@ describe(`Node_Modules`, () => {
           name: `no-deps-bins`,
           version: `1.0.0`,
         });
+      },
+    ),
+  );
+
+  test(`should not hoist package peer dependent on parent if conflict exists`,
+    // . -> dep -> conflict@2 -> unhoistable --> conflict
+    //   -> conflict@1
+    // `unhoistable` cannot be hoisted to the top, otherwise it will use wrong `conflict` version
+    makeTemporaryEnv(
+      {
+        private: true,
+        dependencies: {
+          dep: `file:./dep`,
+          conflict: `file:./conflict1`,
+        },
+      },
+      async ({path, run, source}) => {
+        await writeFile(npath.toPortablePath(`${path}/.yarnrc.yml`), `
+        nodeLinker: "node-modules"
+      `);
+
+        await writeJson(npath.toPortablePath(`${path}/conflict1/package.json`), {
+          name: `conflict`,
+          version: `1.0.0`,
+        });
+
+        await writeJson(npath.toPortablePath(`${path}/conflict2/package.json`), {
+          name: `conflict`,
+          version: `2.0.0`,
+          dependencies: {
+            unhoistable: `file:../unhoistable`,
+          },
+        });
+
+        await writeJson(npath.toPortablePath(`${path}/dep/package.json`), {
+          name: `dep`,
+          version: `1.0.0`,
+          dependencies: {
+            conflict: `file:../conflict2`,
+          },
+        });
+
+        await writeJson(npath.toPortablePath(`${path}/unhoistable/package.json`), {
+          name: `unhoistable`,
+          version: `1.0.0`,
+          peerDependencies: {
+            conflict: `2.0.0`,
+          },
+        });
+        await writeFile(`${path}/packages/unhoistable/index.js`,
+          `module.exports = require('conflict/package.json').version`);
+
+        await run(`install`);
+
+        expect(await xfs.existsPromise(`${path}/node_modules/unhoistable` as PortablePath)).toBe(false);
       },
     ),
   );

--- a/packages/yarnpkg-pnpify/sources/buildNodeModulesTree.ts
+++ b/packages/yarnpkg-pnpify/sources/buildNodeModulesTree.ts
@@ -191,10 +191,6 @@ const buildPackageTree = (pnp: PnpApi, options: NodeModulesTreeOptions): Hoister
           if (depPkg === null)
             throw new Error(`Assertion failed: Expected the package to have been registered`);
 
-          // Skip package self-references
-          if (depLocator.name === locator.name && depLocator.reference === locator.reference)
-            continue;
-
           addPackageToTree(name, depPkg, depLocator, node, pkg);
         }
       }

--- a/packages/yarnpkg-pnpify/sources/hoist.ts
+++ b/packages/yarnpkg-pnpify/sources/hoist.ts
@@ -215,7 +215,7 @@ const hoistTo = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePath:
 
   const hoistIdentMap = getHoistIdentMap(rootNode, popularityMap);
 
-  const hoistIdents = new Set(Array.from(hoistIdentMap.values()).map(x => x[0]));
+  const hoistIdents = new Map(Array.from(hoistIdentMap.entries()).map(([k, v]) => [k, v[0]]));
 
   const hoistedDependencies = rootNode === tree ? new Map() : getHoistedDependencies(rootNode);
 
@@ -225,9 +225,9 @@ const hoistTo = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePath:
     wasStateChanged = false;
     for (const [name, idents] of hoistIdentMap) {
       if (idents.length > 1 && !rootNode.dependencies.has(name)) {
-        hoistIdents.delete(idents[0]);
+        hoistIdents.delete(name);
         idents.shift();
-        hoistIdents.add(idents[0]);
+        hoistIdents.set(name, idents[0]);
         wasStateChanged = true;
       }
     }
@@ -286,7 +286,7 @@ const getSortedReglarDependencies = (node: HoisterWorkTree): Set<HoisterWorkTree
  * @param hoistedDependencies map of dependencies that were hoisted to parent nodes
  * @param hoistIdents idents that should be attempted to be hoisted to the root node
  */
-const hoistGraph = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePath: Set<Locator>, hoistedDependencies: Map<PackageName, HoisterWorkTree>, hoistIdents: Set<Ident>, hoistIdentMap: Map<Ident, Array<Ident>>, options: InternalHoistOptions) => {
+const hoistGraph = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePath: Set<Locator>, hoistedDependencies: Map<PackageName, HoisterWorkTree>, hoistIdents: Map<PackageName, Ident>, hoistIdentMap: Map<Ident, Array<Ident>>, options: InternalHoistOptions) => {
   const seenNodes = new Set<HoisterWorkTree>();
 
   const hoistNode = (nodePath: Array<HoisterWorkTree>, locatorPath: Array<Locator>, node: HoisterWorkTree, newNodes: Set<HoisterWorkTree>) => {
@@ -300,8 +300,8 @@ const hoistGraph = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePa
 
     const parentNode = nodePath[nodePath.length - 1];
     // We cannot hoist self-references
-    const isSelfReference = node.name === parentNode.name;
-    let isHoistable = !isSelfReference && hoistIdents.has(node.ident);
+    const isSelfReference = node.ident === parentNode.ident;
+    let isHoistable = hoistIdents.get(node.name) === node.ident && !isSelfReference;
     if (options.debugLevel >= 2 && !isHoistable)
       reason = `- filled by: ${prettyPrintLocator(hoistIdentMap.get(node.name)![0])} at ${reasonRoot}`;
 

--- a/packages/yarnpkg-pnpify/sources/hoist.ts
+++ b/packages/yarnpkg-pnpify/sources/hoist.ts
@@ -298,7 +298,10 @@ const hoistGraph = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePa
     if (options.debugLevel >= 2)
       reasonRoot = `${Array.from(rootNodePath).map(x => prettyPrintLocator(x)).join(`→`)}`;
 
-    let isHoistable = hoistIdents.has(node.ident);
+    const parentNode = nodePath[nodePath.length - 1];
+    // We cannot hoist self-references
+    const isSelfReference = node.name === parentNode.name;
+    let isHoistable = !isSelfReference && hoistIdents.has(node.ident);
     if (options.debugLevel >= 2 && !isHoistable)
       reason = `- filled by: ${prettyPrintLocator(hoistIdentMap.get(node.name)![0])} at ${reasonRoot}`;
 
@@ -350,7 +353,6 @@ const hoistGraph = (tree: HoisterWorkTree, rootNode: HoisterWorkTree, rootNodePa
     }
 
     if (isHoistable) {
-      const parentNode = nodePath[nodePath.length - 1];
       parentNode.dependencies.delete(node.name);
       parentNode.hoistedDependencies.set(node.name, node);
       parentNode.reasons.delete(node.name);
@@ -484,7 +486,6 @@ const cloneTree = (tree: HoisterTree): HoisterWorkTree => {
   const seenNodes = new Map<HoisterTree, HoisterWorkTree>([[tree, treeCopy]]);
 
   const addNode = (node: HoisterTree, parentNode: HoisterWorkTree) => {
-    // Skip self-references
     let workNode = seenNodes.get(node);
     const isSeen = !!workNode;
     if (!workNode) {

